rfc-135-updating-routes-in-router.md

```diff
@@ -0,0 +1,44 @@
+# Updating routes in Router
+
+## Summary
+
+This RFC proposes changing the way that routes are updated in Router, moving away from Router API manually updating Router instances and instead allowing Router instances to keep themselves up-to-date via polling MongoDB for changes.
+
+## Problem
+
+Whenever a route needs to be added or updated, Content Store calls into Router API to perform the update. Router API knows the address of every Router instance from a static file that it reads at start-up; it then uses this information to call the `/reload` endpoint on every Route instance to let them know that routes have been updated, so that those instances reload the routes they hold in memory.
+
+As GOV.UK moves to containerise more applications and host them in AWS Elastic Container Service (ECS), Router API will no longer be able to talk to individual instances of Router. We therefore need to rethink how these routes are updated within Router instances and to come up with another way of managing the update process.
+
+## Proposal
+
+Router will keep its in-memory cache of routes up-to-date via a two stage process:
+
+  1. Router will perform a simple query against MongoDB returning the state of the database with the current set of routes; ideally this will be a query which contains a checksum or timestamp of last update which can be used by Router instances to check whether their local copy of routes is up-to-date.
```

Comment:
Do we have any more specifics on how this query might work? I guess we don't want to come up with a solution in the RFC, but I think a little more might be helpful to decide whether the overall idea is a good idea or not. (For example, I'm imagining we might need to think a little bit about database indexes or how long it would take to generate said checksum, etc).

Reply:
I wasn't sure how far to go into detail here, as like you say, we haven't done any real research into this and I imagine it'll be something we look at when it comes to implementation. If we're calling a method such as `db.stats()` (or using some other method which returns aggregates) then we might not need to worry about manually creating indexes etc as the queries _should_ already be performant.

On the note of how long it takes to query, so long as the queries return results in a short (and reasonable) amount of time I feel we will be fine.

Reply:
Have added an additional paragraph about this, let us know your thoughts.

Reply:
A couple more ideas to demonstrate that the approach is feasible: [collection.dbHash](https://docs.mongodb.com/v2.6/reference/command/dbHash), or Router API could write a timestamp when the routes collection changes and not use a hash.

Reply:
(Again, sorry for thread necromancy but seemed like the most relevant place to continue the discussion.)

To echo Tom's point, I think the details of the polling mechanism are quite crucial here and to defer their consideration until later as a mere implementation detail would be a mistake. The concept may be simple, but how Router API denotes that there's been a change is a different matter and I think it reduces to a non-trivial classical concurrency problem.

For example, having a "last updated" timestamp on the database would be super convenient for implementing the polling loop in Router, but how would Router API update that timestamp? Just write to a specific record in the database each time a change happens? That means all writes will contend for the same lock. Maybe that's fine, say if we're only writing at most a handful of times a second. How frequently do updates happen in the worst case? (Are there bulk republishing operations which generate a lot of changes all at once, for example?)

I'm pretty sure hashing is a red herring - reading through the entire dataset on every write to compute the hash would be infeasible, and doing something incremental involves fancy techniques which Mongo almost certainly doesn't implement. I don't think it's necessary either - a counter or a timestamp would do just fine (and even that is challenging enough!)

There are a couple of things in our favour. Firstly, Mongo already implements replication, so we may be able to leverage that (for example perhaps the replication counters are suitable to detecting when there's been an update). Secondly, we only need eventual consistency and we don't care about linearizability (I'm fairly sure we don't, anyway).

Reply:
(tl;dr: let's figure out if we can get away with just polling `rs.status()`)

Reply:
For the updating of a timestamp as the mechanism for Router instances to detect changes, I would hope that Mongo can actually take care of this for us and what we then don't need to worry about Router API instances contending for the same lock; we just use the timestamp returned either in a call to one of the above mentioned methods. If I'm being naive here then please let me know!

Reply:
Another option we have is to use the primary machine's `oplog`. [From the docs](https://docs.mongodb.com/manual/core/replica-set-oplog/)...

- _The oplog (operations log) is a special capped collection that keeps a rolling record of all operations that modify the data stored in your databases._

- _MongoDB applies database operations on the primary and then records the operations on the primary’s oplog. The secondary members then copy and apply these operations in an asynchronous process. All replica set members contain a copy of the oplog, in the local.oplog.rs collection, which allows them to maintain the current state of the database._

- _Each operation in the oplog is idempotent._

From the oplog, we can call the method `rs.printReplicationInfo()` to get the timestamp of the last update to the primary which can then be used to poll for changes in Router instances.

Example output of `rs.printReplicationInfo()` on the Router database:
```
rs.printReplicationInfo()
configured oplog size:   990MB
log length start to end: 130748secs (36.32hrs)
oplog first event time:  Mon Mar 08 2021 03:24:12 GMT+0000 (UTC)
oplog last event time:   Tue Mar 09 2021 15:43:20 GMT+0000 (UTC)
now:                     Tue Mar 09 2021 16:21:54 GMT+0000 (UTC)
``` 

Importantly, as the oplog is a log of all operations (modifications) performed on the primary, there is no need to perform a table scan or other high-workload operation on the primary.

Reply:
Aha, so I was talking nonsense about the 1-second precision being a problem. We can just compare Timestamp objects from `rs.status()` and we should be fine?

Reply:
Just to clarify what's being said here, we could use the `optime` timestamp that's returned when calling `rs.status()`, which is the internal MongoDB BSON timestamp. Importantly though, it _does_ give us the actual timestamp and the ordinal which together form the most recent update to a given database instance, whether the primary or one of the secondaries. The format it returns this in is:

```
...
"optime" : Timestamp(1615383267, 4),
...
```

We are given this timestamp for all members of the replica set, so as long as we're consistent about which replica we're reading from, Router instances should be able to use this value when polling to determine whether there are any new route updates to fetch and if so, to reload it's routes from MongoDB. This timestamp gives us confidence that there will be no race conditions of the kind described previously, so there's also no need to do any force refreshes of routes in Router.